### DISCUSSION TITLE
When will the 5.x version be released to support jetty12?

### DISCUSSION THREAD

**Discussion by stevenniu9527** [dropwizard/dropwizard#8875] [Development]
Hi, When will the 5.x version be released to support jetty12?
and Is there an migration instructions？

**Comment by joschi** [comment]
There is currently no release date for Dropwizard 5.x and there's currently not a lot of active work being done on that branch.

**Reply by mattnelson** [reply]
The jetty version used in 3.x and 4.x will be End of Community Support on Jan 1st, 2024. A bit less than 6 months from now.

https://github.com/jetty/jetty.project/issues/10485

**Reply by joschi** [reply]
> End of Community Support on Jan 1st, 2024

That's hopefully Jan 1st, **2025**. 😉

**Reply by zUniQueX** [reply]
I already wanted to create the first beta for 5.x but have been quite busy over the last weeks. Hopefully, I'll merge the pending Jetty update next week, then we could release the first beta IMHO. Waiting for Jackson 3.x without an ETA probably won't be possible for Dropwizard 5 then.

**Reply by mattnelson** [reply]
> Waiting for Jackson 3.x without an ETA probably won't be possible for Dropwizard 5 then.

Does this mean dropwizard 5 would stay with jackson 2?
or that an ETA for dropwizard 5 is not possible without an ETA for jackson 3?

**Reply by zUniQueX** [reply]
If the beta is released with Jackson 2 this week (or maybe next week) and we don't receive much negative feedback, the release of Dropwizard 5 could be quite soon. That means without an ETA for Jackson 3 it's unlikely to make it into Dropwizard 5.

**Reply by sleberknight** [reply]
> > End of Community Support on Jan 1st, 2024
> 
> That's hopefully Jan 1st, **2025**. 😉

I also misread this and thought it was upcoming in January 1, **2025**.

But, the Jetty downloads page says:

> January 1st, **2024** marked the end of community support for the Jetty 10 and 11 releases.

-- https://jetty.org/download.html

I admit I do not know exactly what "end of community support" means. Jetty 10.x and 11.x continue to get patch updates. Versions 10.0.22 and 11.0.22 were both released a month ago on July 4, 2024 according to mvnrepository.com. We continue to update to the latest 11.x patch releases as they come out, and things still work fine.

**Reply by Athou** [reply]
From what I understand from https://github.com/jetty/jetty.project/issues/10485, Jetty 10 and Jetty 11 currently only receive security patches, and will be completely unsupported by the end of the year.

> Once EoCS is reached, users of these older versions will continue to see releases for security and critical issues until they go full End of Life (EOL), and the only ongoing support provided will be for Webtide customers. By way of example, Jetty 9 EoCS was announced in May of 2022, and there have been five customer-sponsored releases since that announcement.

> As for the official EOL of Jetty 10 and Jetty 11, we do not foresee this before 1 January 2025; however, that may change based on community migration uptake for Jetty 12.

**Reply by nabeelq** [reply]
> I already wanted to create the first beta for 5.x but have been quite busy over the last weeks. Hopefully, I'll merge the pending Jetty update next week, then we could release the first beta IMHO. Waiting for Jackson 3.x without an ETA probably won't be possible for Dropwizard 5 then.

any updates on it ?

**Reply by clarkbreyman** [reply]
@joschi  - Per Joakim (https://github.com/jetty/jetty.project/pull/12012#issuecomment-2416450253), the patch for CVE-2024-6763 won't be backported to 11. It looks like the current release/5.x has the patch in it. 

Do you feel that the current alpha is stable? What additionally work is needed to bless 5.x?

**Reply by robbytx** [reply]
> Do you feel that the current alpha is stable? What additionally work is needed to bless 5.x?

I'd love to know more about this. Are there any known outstanding issues?

From a July comment above:

> If the beta is released with Jackson 2 this week (or maybe next week) and we don't receive much negative feedback, the release of Dropwizard 5 could be quite soon.

If there are no outstanding issues, and it's just waiting on more testing from the community, then releasing a "beta" version could be helpful.  IME DropWizard has been fairly stable over the years -- since it's built on top of other quite stable components -- but the "alpha" label suggests to me that there are still several things needing to be worked out.

If there are known outstanding issues, then it would be helpful to understand what those are, so folks can make an informed decision about upgrading to a pre-release 5.x vs staying on a known-vulnerable Jetty version.

**Reply by sbonebrake** [reply]
Circling back on this as we are now at EOL for Jetty 10 and 11. Would be interested if @joschi has any thoughts the stability of Dropwizard 5.x. Perhaps it's ready for a Beta?

**Reply by joschi** [reply]
@sbonebrake If we just retagged the latest alpha as beta, would that work for you? 😉 

In all seriousness: Try upgrading your application to the latest Dropwizard 5.0.x alpha release and let us know how it went and what you would expect in the migration notes.

The more people do this, the quicker we can release Dropwizard 5.0.0.

**Reply by Mahoney** [reply]
FWIW I just upgraded our (very mature - in prod for 9 years) codebase to `5.0.0-alpha.5`. We have 2,211 tests, the vast majority of which are full stack using the `DropwizardAppRule` against HSQLDB, and they all pass OK.

In the production code the changes were minimal:
```java
import org.eclipse.jetty.server.session.SessionHandler;
import org.eclipse.jetty.servlets.CrossOriginFilter;
```
became
```java
import org.eclipse.jetty.ee10.servlet.SessionHandler;
import org.eclipse.jetty.ee10.servlets.CrossOriginFilter;
```

In `build.gradle` I had more changes:
```groovy
dependencies {
    implementation platform("org.eclipse.jetty:jetty-bom:11.0.24")
    implementation "org.eclipse.jetty:jetty-servlet"
    implementation "org.eclipse.jetty:jetty-servlets"
    implementation 'org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2'
}
```
became
```
dependencies {
    implementation platform("org.eclipse.jetty:jetty-bom:12.0.16")
    implementation platform("org.eclipse.jetty.ee10:jetty-ee10-bom:12.0.16")
    implementation "org.eclipse.jetty.ee10:jetty-ee10-servlet"
    implementation "org.eclipse.jetty.ee10:jetty-ee10-servlets"
    implementation 'jakarta.servlet:jakarta.servlet-api:6.1.0'
}
```

I had more issues with my tests, because I was depending on `io.dropwizard.modules:dropwizard-testing-junit4` and the `DropwizardAppRule`. I actually just ended up copying that class into my own codebase, switching to `DropwizardAppExtension` was going to be too painful.

No decision yet made as to whether we risk it in live.

**Reply by dermot-hardy** [reply]
> **JS:** Try upgrading your application to the latest Dropwizard 5.0.x alpha release and let us know how it went...

We upgraded to `5.0.0-alpha.4` back in November due to an internal company mandate to stop using Jetty 11 given its EOL status.  The main issue we encountered was the #9570 issue, which was subsequently resolved in `5.0.0-alpha.5`.  Here is an example PR of the changes made in case that is helpful: [JobService/job-service#458](https://github.com/JobService/job-service/pull/458).

We deployed a series of services using `5.0.0-alpha.5` to production in December and there have been no issues to date.

**Reply by josephlbarnett** [reply]
fwiw it looks like jetty somewhat significantly changed their websocket apis enough that the way I was using them is not quite as straightforward to match -- will spend some time on it but at least the non-websocket aspects did port over very easily (just pom/package changes)

**Reply by josephlbarnett** [reply]
ok example of changes needed that seem to be mostly working but still validating here: https://github.com/josephlbarnett/leakycauldron/compare/main...dw5-jetty-12-wip
highlights:
 * ws api changes (WebSocketAdapter -> Session.Listener.AutoDemanding, additional callback params/straightforward API changes, weird WS setup to initiate from a jersey resource instead of a top level servlet has changed in every recent major jetty version...)
 * testing/mock changes, especially Request/Response variants
 * movement from deprecated CrossOriginFilter to CrossOriginHandler and trying to insert that above the DW MutableServletContextHandler via env.applicationContext.insertHandler()
